### ISSUE TITLE
questions about sequence model with RNN

### ISSUE THREAD

**Issue by JH-Lam** [pandalabme/d2l#1]
Hi thank you for opening this project. Now I have a couple of questions in 'sequence implemented by RNN', ie. https://github.com/pandalabme/d2l/blob/main/exercises/ch9/9_6_4_Exercises.md :
1. do you think 'a too complex model' approximately equals  ' Insufficient training data' , and vice versa 
2. is this code block complete? as I saw hidden state and batch size are never used, it seems like common linear regression 

expect your reply ~

**Comment by pandalabme** [comment]
> Hi thank you for opening this project. Now I have a couple of questions in 'sequence implemented by RNN', ie. https://github.com/pandalabme/d2l/blob/main/exercises/ch9/9_6_4_Exercises.md :
> 
> 1. do you think 'a too complex model' approximately equals  ' Insufficient training data' , and vice versa
> 2. is this code block complete? as I saw hidden state and batch size are never used, it seems like common linear regression
> 
> expect your reply ~

Sorry for my late reply, and thank you for pointing out my mistake! I am a newbie to deep learning, my answers may not be right. My opinions about the questions are below:
**Q1: do you think 'a too complex model' approximately equals ' Insufficient training data' , and vice versa**

**A1： In my opinion , "A too complex model" and "insufficient training data" are related concepts in the context of machine learning, but they are not exactly equivalent, and they are not the same thing. They represent different aspects of the machine learning process:

1. **A Too Complex Model:** This typically refers to a model that has too many parameters or is overly intricate, leading to overfitting. Overfitting occurs when a model learns to fit the training data extremely well but fails to generalize to new, unseen data. Complex models can capture noise in the training data, making them less useful for real-world applications. Simplifying the model by reducing its complexity, such as by reducing the number of layers in a neural network, can help mitigate overfitting.

2. **Insufficient Training Data:** This refers to the problem of not having enough diverse and representative data to train a machine learning model effectively. When you have insufficient data, it can be challenging for the model to learn meaningful patterns and relationships in the data. In such cases, the model might not perform well, even if it's not overly complex.

While these two concepts are related in the sense that both can lead to poor model performance, they are distinct issues. It's possible to have a simple model with insufficient data or a complex model with insufficient data. In either case, the model may not perform well because it lacks the necessary information to make accurate predictions.

Ultimately, the goal in machine learning is to strike a balance between model complexity and data availability to build a model that generalizes well to unseen data.**

**Q2: is this code block complete? as I saw hidden state and batch size are never used, it seems like common linear regression**

**A2: Yes, you are right, there is something wrong with the origin code, I made a mistake in generating dataset for the rnn model, the updated code is [here](https://github.com/pandalabme/d2l/blob/main/exercises/ch9/9_6_4_Exercises.md)**

I am very happy to have the opportunity to discuss issues together. Please feel free to ask any questions!

**Comment by JH-Lam** [comment]
I got it , thank you, 
q1
both cases are not same while they seem relative to other

q2
I just go into your code blocks. I described 'the state is never used' mean the 'state never reused in rnn()':
`def forward(self, X, state=None):
        rnn_outputs, _ = self.rnn(X, state) 
        return self.linear(rnn_outputs[-1])`
should be changed to:
`def forward(self, X, state=None):
        rnn_outputs, state = self.rnn(X, state)
        return self.linear(rnn_outputs[-1]), state`
 but may be i was wrong since I used CN version source to do a sequence-RNN.ipynb while you are using newer en version. the latter is changed a lot in both text and code style. and my result is much more close to sequnce.ipynb than yours, and a bit more smooth than official linear sequence.ipynb.

q3
what's your progress in deepleanring? I see you learned many advanced techniques ,say attention, transformer etc in your blogs site which I dont learn yet.
which domain in deeplearning are you focus, images or NLP?
ps: both your english and DL techniques are good, I learned much from your git, thanks again.

**Comment by pandalabme** [comment]
a2
Yes, there are a lot of differences between [EN](https://d2l.ai/) and [CN](https://zh.d2l.ai/) version. I prefer the EN version since it has more details.

But I am a little confused about the idea that 'state never reused in rnn()'. 

In the code  ```def forward(self, X, state=None): rnn_outputs, _ = self.rnn(X, state)  return self.linear(rnn_outputs[-1])``` , the dimension of input ```X``` is (seq_len,batch_size,input_size), and dimension of output ```rnn_outputs``` is (seq_len,batch_size,hidden_size), meanwhile input ```state``` is the initiated hidden state **h0**, and output ```_``` is the last hidden state **ht**, which is the same with ```rnn_outputs[-1]```

In the first time step, ```torch.nn.RNN``` computes **h1** by the input ```X[0,:,:]``` and **h0**, and saved in ```rnn_outputs[0,:,:]```. During the next steps, ```torch.nn.RNN``` will computes **i+1**th hidden state by ```X[i,:,:]``` and **hi** (which is ```rnn_outputs[i-1,:,:]```)  automatically. 

a3
I read [d2l](https://d2l.ai/) chapter by chapter, and I'm on chapter 15 now. 
I think both domains are very interesting, while NLP catches my eye more. When I finish d2l, I will read some classic papers in NLP and write some findings in my blog.

Thanks for your encouragement! Hoping my learning experience can help others is the original intention behind starting my blog.

**Comment by JH-Lam** [comment]
Q2:  but you never reused the state derived in last batch into next batch , right? that is you returned only the linear(output[-1]) but without old state for next iteration which did in official. **excepts the state itself will be changed somewhere.**

Q3: what kind of your job about now, related to AI?

**Comment by JH-Lam** [comment]
in case Q2: 
it maybe cause by different code writing style:
**-for RNNLMScratch , it returns w/o state**
![Xnip2023-10-16_11-53-24](https://github.com/pandalabme/d2l/assets/53388447/10b63349-88ee-45da-9d80-246cb07a6166)

**-but for RNNScratch , it does**
![Xnip2023-10-16_12-05-56](https://github.com/pandalabme/d2l/assets/53388447/7d3d6602-ef28-46c7-855c-0e443e84c6a3)

**Comment by pandalabme** [comment]
RNNLMScratch and RNNScratch are different models, RNNScratch is a RNN model implemented from scratch (just like  ```torch.nn.RNN```). While RNNLMScratch is a RNN-based language model, which just need outputs of each time step produced by RNN, and the last state helps little, so the code use ```_``` to represent.

**Comment by pandalabme** [comment]
A3
I was a data developer worked for an Internet company, hoping to find some opportunities within AI in the future.

**Comment by JH-Lam** [comment]
got it. 
for my test cases, it decreases loss a bit more in case sequence -rnn.ipynb while improves ppl at least 0.3 in case rnn-scratch.ipynb if reused state for next batch.

good luck to new job~

**Comment by pandalabme** [comment]
I think the reason why reused state has better result is that the ```train_iter``` generated by [load_data_time_machine](https://github.com/d2l-ai/d2l-zh/blob/master/d2l/torch.py#L685) is not shuffle(```use_random_iter=False```). So the reused state can give more information outside the time step, and I guess that the result might be worse if the dataset was shuffle( the ppl might not be improved 0.3).

Looking forward to your test case sharing and thanks for your wishes!
